File: src/Algorithm/Maze/MouseLocal.java
package src.Algorithm.Maze;

import java.util.ArrayList;
import java.util.Arrays;

import src.Constants;
import src.API.API;

public class MouseLocal {
    private Cell[][] mazeCells;
    private int[] mousePosition;
    private int[] mouseDirection;

    public MouseLocal() {
        mazeCells = new Cell[Constants.MazeConstants.numRows][Constants.MazeConstants.numCols];
        mousePosition = Constants.MouseConstants.startingMousePosition;
        mouseDirection = Constants.MouseConstants.startingMouseDirection;

        /*
         * Instantiates each cell in the Maze with an x and y coordinate.
         * 
         * It may be beneficial to visualize the array as the following:
         * - [column 0, column 1, column 2, ..., column 15]
         * - with each column: [row 0 elem, row 1 elem, row 2 elem, ..., row 15 elem]
         * 
         * 0, 0 signifies the bottom-left cell. 15, 15 signifies the top-right cell.
         * Thus, getting element 0, 1 will return the cell above the bottom-left cell
         * according to the following array logic.
         */
        setUpMazeLocal();
    }

    /**
     * Function to set up the maze with cells marked by their x, y coordinates.
     * Maze is characterized by:
     * 
     * ^ 15,15
     * |
     * +y
     * 0,0 +x -->
     */
    private void setUpMazeLocal() {
        for (int i = 0; i < Constants.MazeConstants.numCols; i++) { // X-direction
            for (int j = 0; j < Constants.MazeConstants.numRows; j++) { // Y-direction
                mazeCells[i][j] = new Cell(i, j);
            }
        }
    }

    /**
     * Adjusts the mouse's locally held direction based on the turn performed.
     * 
     * @param halfStepsLeft  # of 45 degree turns to the left.
     * @param halfStepsRight # of 45 degree turns to the right.
     */
    public void turnMouseLocal(int halfStepsLeft, int halfStepsRight) {
        try {
            int[][] possibleMouseDirections = Constants.MouseConstants.possibleMouseDirections;
            mouseDirection = possibleMouseDirections[(findDirectionIndexInPossibleDirections(mouseDirection)
                    + halfStepsRight - halfStepsLeft + possibleMouseDirections.length)
                    % possibleMouseDirections.length];
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        }
    }

    /**
     * Adjusts the mouse's locally held direction based on a desired direction.
     * 
     * @param newDirection The new direction to turn the mouse to.
     */
    public int[] turnMouseLocal(int[] newDirection) {
        int[] halfStepCount = obtainHalfStepCount(newDirection);
        turnMouseLocal((halfStepCount[1] == -1) ? halfStepCount[0] : 0, (halfStepCount[1] == 1) ? halfStepCount[0] : 0);
        return halfStepCount;
    }

    /**
     * Obtains optimal # of half steps (left/right) to reach a desired position
     * 
     * @param newDirection The desired direction to turn the mouse to.
     * @return An array containing the # of half steps and the direction to turn
     *         (left or right).
     */
    public int[] obtainHalfStepCount(int[] newDirection) {
        int halfStepsRight = (Constants.MouseConstants.possibleMouseDirections.length
                + findDirectionIndexInPossibleDirections(newDirection)
                - findDirectionIndexInPossibleDirections(mouseDirection))
                % Constants.MouseConstants.possibleMouseDirections.length;
        int halfStepsLeft = (Constants.MouseConstants.possibleMouseDirections.length
                + findDirectionIndexInPossibleDirections(mouseDirection)
                - findDirectionIndexInPossibleDirections(newDirection))
                % Constants.MouseConstants.possibleMouseDirections.length;

        // System.err.println("Half steps right: " + halfStepsRight + " Half steps left:
        // " + halfStepsLeft);
        int halfSteps = Math.min(halfStepsRight, halfStepsLeft);

        int direction = (halfStepsRight < halfStepsLeft) ? 1 : -1; // 1 signifies right, -1 signifies left.

        return new int[] { Math.abs(halfSteps), direction };

    }

    /**
     * Finds the index of a given direction in the array of possible mouse
     * directions.
     * 
     * @param direction The direction fed in.
     * @return The index of the direction in the possibleMouseDirection array.
     * @throws IllegalArgumentException Thrown if the direction is not listed in as
     *                                  a possible direction.
     */
    private int findDirectionIndexInPossibleDirections(int[] direction) throws IllegalArgumentException {
        int[][] possibleMouseDirections = Constants.MouseConstants.possibleMouseDirections;
        for (int i = 0; i < possibleMouseDirections.length; i++) {
            if (Arrays.equals(possibleMouseDirections[i], direction)) {
                return i;
            }
        }
        throw new IllegalArgumentException("Direction not listed as a possible mouse direction: " + direction);
    }

    /**
     * Moves the mouse forward in the direction it is currently in.
     */
    public void moveForwardLocal() {
        int newXPosition = mousePosition[0] + mouseDirection[0];
        int newYPosition = mousePosition[1] + mouseDirection[1];

        if (isValidCell(newXPosition, newYPosition)) {
            mousePosition = new int[] { newXPosition, newYPosition };
        } else {
            System.err.println(
                    "Invalid position (ack), mouse cannot move to (" + newXPosition + "," + newYPosition + ")");
        }
    }

    /**
     * Adds a wall to the local maze based on a direction fed in.
     * 
     * @param x         The x-position of the cell to add a wall to.
     * @param y         The y-position of the cell to add a wall to.
     * @param direction The direction to add the wall to. Could be n, ne, e, se, s,
     *                  sw, w, nw.
     */
    public void addWallLocal(int x, int y, int[] direction) {
        int neighboringCellX = x + direction[0];
        int neighboringCellY = y + direction[1];

        if (isValidCell(neighboringCellX, neighboringCellY)) {
            mazeCells[x][y].addWall(direction, true);
            mazeCells[neighboringCellX][neighboringCellY].addWall(new int[] { -direction[0], -direction[1] }, true);
            // System.err.println("Shared wall cell found :)");
        } else {
            mazeCells[x][y].addWall(direction, false);
            // System.err.println("Edge cell found :)"); // FIXME: Remove later.
        }
    }

    /**
     * Returns if the cell is in the constricted space of the maze.
     * 
     * @param x The x-position of the cell.
     * @param y The y-position of the cell.
     * @return If the cell is in the constricted space of the maze.
     */
    public boolean isValidCell(int x, int y) {
        return x >= 0 && x < Constants.MazeConstants.numCols && y >= 0 && y < Constants.MazeConstants.numRows;
    }

    /**
     * Returns a boolean stating whether the mouse can move between two cells.
     * 
     * @param cell1 The cell the mouse is moving from.
     * @param cell2 The cell the mouse is moving to.
     * @return If the mouse can move between the two cells.
     */
    public Movement getMovement(Cell cell1, Cell cell2, boolean diagonalsAllowed) {
        int[] direction = new int[] { cell2.getX() - cell1.getX(), cell2.getY() - cell1.getY() };
        try {
            if(Math.sqrt(Math.pow(direction[0], 2) + Math.pow(direction[1], 2)) != 1 && Math.sqrt(Math.pow(direction[0], 2) + Math.pow(direction[1], 2)) != Math.sqrt(2)) {
                return new Movement(false, direction);
            }
            // Cardinal direction movement.
            boolean canMove = !cell1.getWallExists(direction);
            Movement cardinalMovement = new Movement(canMove, direction);

            return cardinalMovement;
        } catch (IllegalArgumentException e) {
            Movement diagonalMovement = new Movement(false, direction);
            if (!diagonalsAllowed || !cell2.getIsExplored()) {
                return diagonalMovement;
            }

            // System.err.println("Current cell: " + cell1.getX() + ", " + cell1.getY() + " direction move: " + direction[0] + ", " + direction[1] + " to cell: " + cell2.getX() + ", " + cell2.getY());

            // If part of the eight cardinal directions, but not the four cardinal directions, check the following:
            boolean diagonalVertical = false;
            boolean diagonalHorizontal = false;
            // E.g. direction is (-1, -1) --> check new cell's north and east walls as well as current cell's south and west walls.
            // In this example, if either the new cell's north + current cell's west (upper diagonal) don't exist OR the new cell's east + current cell's south (lower diagonal) don't exist, the mouse can move diagonally.
            int[] horizontalDirectionCheck = new int[] {direction[0], 0}; // E.g. -1 0
            int[] verticalDirectionCheck = new int[] {0, direction[1]}; // E.g. 0 -1

            // System.err.println("Direction: " + direction[0] + ", " + direction[1]);

            diagonalHorizontal = (!cell1.getWallExists(horizontalDirectionCheck) && !cell2.getWallExists(new int[] {0, -verticalDirectionCheck[1]}));
            diagonalVertical = (!cell1.getWallExists(verticalDirectionCheck) && !cell2.getWallExists(new int[] {-horizontalDirectionCheck[0], 0}));

            diagonalMovement.setCanMove(diagonalHorizontal || diagonalVertical);
            if (!diagonalMovement.getCanMove()) {
                return diagonalMovement;
            }

            // Classifies if the horizontal diagonal is the left or right diagonal of the mouse.
            boolean isHorizontalRight = (findDirectionIndexInPossibleDirections(verticalDirectionCheck) < findDirectionIndexInPossibleDirections(horizontalDirectionCheck));
            boolean isVerticalRight = !isHorizontalRight;

            if (isHorizontalRight && diagonalHorizontal) {
                diagonalMovement.setIsDiagonal(true);
                diagonalMovement.setLeftOrRightDiagonal("right");
                diagonalMovement.setCellToMoveToFirst(getCell(cell1.getX() + horizontalDirectionCheck[0], cell1.getY()));
            } else if (!isHorizontalRight && diagonalHorizontal) {
                diagonalMovement.setIsDiagonal(true);
                diagonalMovement.setLeftOrRightDiagonal("left");
                diagonalMovement.setCellToMoveToFirst(getCell(cell1.getX() + horizontalDirectionCheck[0], cell1.getY()));
            } else if (isVerticalRight && diagonalVertical) {
                diagonalMovement.setIsDiagonal(true);
                diagonalMovement.setLeftOrRightDiagonal("right");
                diagonalMovement.setCellToMoveToFirst(getCell(cell1.getX(), cell1.getY() + verticalDirectionCheck[1]));
            } else {
                diagonalMovement.setIsDiagonal(true);
                diagonalMovement.setLeftOrRightDiagonal("left");
                diagonalMovement.setCellToMoveToFirst(getCell(cell1.getX(), cell1.getY() + verticalDirectionCheck[1]));
            }

            return diagonalMovement;
        }
    }

    /**
     * Returns the direction offsets (an array of two integers) based on a given
     * direction.
     * 
     * @param direction The direction to get the offset for represented as a string
     *                  (n, ne, e, se, s, sw, w, nw).
     * @return The direction offsets based on the given direction.
     */
    public int[] getDirectionOffset(String direction) {
        String[] possibleDirections = new String[] { "n", "ne", "e", "se", "s", "sw", "w", "nw" };
        return Constants.MouseConstants.possibleMouseDirections[Arrays.asList(possibleDirections).indexOf(direction)];
    }

    /**
     * Returns the direction as a string (n, ne, e, se, s, sw, w, nw) based on a
     * given direction.
     * 
     * @param direction The direction as offset {+x, +y} to get the direction for.
     * @return The direction as a string (n, ne, e, se, s, sw, w, nw).
     */
    public String getDirectionAsString(int[] direction) {
        String[] possibleDirections = new String[] { "n", "ne", "e", "se", "s", "sw", "w", "nw" };
        return possibleDirections[findDirectionIndexInPossibleDirections(direction)];
    }

    /**
     * Returns the direction to the left of the current mouse's direction.
     * 
     * @return The direction to the left of the current mouse's direction.
     */
    public String getDirectionToTheLeft() {
        int[] newDirection = Constants.MouseConstants.possibleMouseDirections[(findDirectionIndexInPossibleDirections(
                mouseDirection) + 6) % 8];
        return getDirectionAsString(newDirection);
    }

    /**
     * Returns the direction to the right of the current mouse's direction.
     * 
     * @return The direction to the right of the current mouse's direction.
     */
    public String getDirectionToTheRight() {
        int[] newDirection = Constants.MouseConstants.possibleMouseDirections[(findDirectionIndexInPossibleDirections(
                mouseDirection) + 2) % 8];
        return getDirectionAsString(newDirection);
    }

    /**
     * Returns the mouse's current direction as an offset.
     * 
     * @return The mouse's current direction as an offset.
     */
    public int[] getMouseDirection() {
        return mouseDirection;
    }

    /**
     * Returns the maze with all cells.
     * 
     * @return The maze with all cells.
     */
    public Cell[][] getMazeCells() {
        return mazeCells;
    }

    /**
     * Returns a particular cell in the maze at position x,y.
     * 
     * @param x The x-position of the cell.
     * @param y The y-position of the cell.
     * @return The cell at position x,y in the maze.
     */
    public Cell getCell(int x, int y) {
        return mazeCells[x][y];
    }

    /**
     * Returns the mouse's current position in the maze.
     * 
     * @return The mouse's current position in the maze.
     */
    public Cell getMousePosition() {
        return getCell(mousePosition[0], mousePosition[1]);
    }

    /**
     * Returns a 2D representation of the maze as a string (formatted as a map).
     * 
     * @return A 2D representation of the maze as a string (formatted as a map).
     */
    public String localMazeToString() {
        StringBuilder mazeString = new StringBuilder("Maze:\n");
        int numRows = Constants.MazeConstants.numRows;
        int numCols = Constants.MazeConstants.numCols;

        // 0 15 --> 15, 15 (Top row of the maze)
        for (int i = 0; i < numCols; i++) {
            mazeString.append("+---");
        }
        mazeString.append("+\n");

        for (int i = numRows - 1; i >= 0; i--) {
            mazeString.append(printRow(i));
            mazeString.append("\n");
        }

        return mazeString.toString();
    }

    /**
     * Prints the walls of a row of cells in a maze.
     * 
     * @param rowNumber The row number to print the walls of the cells of.
     * @return The walls of the row of cells in the maze formatted as a map.
     */
    private String printRow(int rowNumber) {
        StringBuilder rowString = new StringBuilder("|");
        int numCols = Constants.MazeConstants.numCols;

        for (int i = 0; i < numCols; i++) {
            Cell cellAnalyzed = getCell(i, rowNumber);
            if (cellAnalyzed.getWallExists(new int[] { 1, 0 })) {
                rowString.append("   |");
            } else {
                rowString.append("    ");
            }
        }
        rowString.append("\n");

        for (int i = 0; i < numCols; i++) {
            Cell cellAnalyzed = getCell(i, rowNumber);
            if (cellAnalyzed.getWallExists(new int[] { 0, -1 })) {
                rowString.append("+---");
            } else {
                rowString.append("+   ");
            }
        }
        rowString.append("+\n");

        return rowString.toString();
    }

    /**
     * Gets all valid neighbors in n, e, s, w direcitons.
     * 
     * @param cell The cell to get the neighbors of.
     * @return All valid neighbors in n, e, s, w directions.
     */
    public ArrayList<Cell> getNeighbors(Cell cell, boolean diagonalsAllowed) {
        int x = cell.getX();
        int y = cell.getY();
        ArrayList<Cell> neighbors = new ArrayList<>();

        int possibleDirections[][] = Constants.MouseConstants.possibleMouseDirections;

        for (int i = 0; i < possibleDirections.length; i++) {
            int newX = x + possibleDirections[i][0];
            int newY = y + possibleDirections[i][1];

            if (isValidCell(newX, newY) && (diagonalsAllowed ? true : (possibleDirections[i][0] == 0 || possibleDirections[i][1] == 0))) {
                Cell neighborCell = getCell(newX, newY);
                neighbors.add(neighborCell);
            }
        }
        return neighbors;
    }

    public static boolean isSame(Cell cell1, Cell cell2) {
        return cell1.getX() == cell2.getX() && cell1.getY() == cell2.getY();
    }

    public static double euclideanDistance(Cell cell1, Cell cell2) {
        return Math.sqrt(Math.pow(cell1.getX() - cell2.getX(), 2) + Math.pow(cell1.getY() - cell2.getY(), 2));
    }

    public static double octileDistance(Cell cell1, Cell cell2) {
        int distanceX = Math.abs(cell1.getX() - cell2.getX());
        int distanceY = Math.abs(cell1.getX() - cell2.getX());
        return (distanceX + distanceY) + (Math.sqrt(2) - 2) * Math.min(distanceX, distanceY);
    }

    public void resetCosts() {
        for (int x = 0; x < Constants.MazeConstants.numCols; x++) {
            for (int y = 0; y < Constants.MazeConstants.numRows; y++) {
                Cell cell = getCell(x, y);
                cell.setCostFromStart(Double.POSITIVE_INFINITY);
                cell.setTotalCost(Double.POSITIVE_INFINITY);
            }
        }
    }

    /**
     * Detects and sets walls in front, left, and right of the mouse.
     * 
     * @param mouse The mouse object.
     * @param API   The API object.
     */
    public void detectAndSetWalls(API api) {
        Cell currCell = getMousePosition();
        if (api.wallFront()) {
            api.setWall(currCell.getX(), currCell.getY(), getDirectionAsString(getMouseDirection()));
        }
        if (api.wallLeft()) {
            api.setWall(currCell.getX(), currCell.getY(), getDirectionToTheLeft());
        }
        if (api.wallRight()) {
            api.setWall(currCell.getX(), currCell.getY(), getDirectionToTheRight());
        }
    }

    public ArrayList<Cell> getGoalCells() {
        ArrayList<int[]> goalPoses = Constants.MazeConstants.getGoalCells();
        ArrayList<Cell> goalCells = new ArrayList<Cell>();

        for (int[] goalPos : goalPoses) {
            goalCells.add(getCell(goalPos[0], goalPos[1]));
        }
        return goalCells;
    }


    /**
     * Check if a cell is one of our four avoided goals.
     */
    public boolean isGoalCell(Cell cell, ArrayList<Cell> goalCells) {
        boolean isGoal = false;
        for (Cell goal : goalCells) {
            isGoal = isGoal || isSame(cell, goal);
        }
        return isGoal;
    }
}

File: src/Algorithm/Maze/MouseLocal.cpp
#include "MouseLocal.h"
#include "../../Constants.h"
#include <iostream>
#include <sstream>
#include <cmath>
#include <algorithm>

// Constructor
MouseLocal::MouseLocal() {
    // Initialize mazeCells with numRows x numCols
    mazeCells.resize(Constants::MazeConstants::numRows, std::vector<Cell*>(Constants::MazeConstants::numCols, nullptr));
    
    // Initialize cells with their coordinates
    setUpMazeLocal();

    // Initialize mousePosition and mouseDirection from Constants
    mousePosition = { Constants::MouseConstants::startingMousePosition[0], Constants::MouseConstants::startingMousePosition[1] };
    mouseDirection = { Constants::MouseConstants::startingMouseDirection[0], Constants::MouseConstants::startingMouseDirection[1] };
}

// Sets up the maze with cells
void MouseLocal::setUpMazeLocal() {
    for(int i = 0; i < Constants::MazeConstants::numCols; ++i) { // X-direction
        for(int j = 0; j < Constants::MazeConstants::numRows; ++j) { // Y-direction
            mazeCells[j][i] = new Cell(i, j); // Assuming mazeCells[row][col], row=j, col=i
        }
    }
}

void MouseLocal::deleteMazeLocal()
{
    for(int i = 0; i < Constants::MazeConstants::numCols; ++i) { // X-direction
        for(int j = 0; j < Constants::MazeConstants::numRows; ++j) { // Y-direction
            delete mazeCells[j][i];
        }
    }
}

// Adjusts the mouse's direction based on turn steps
void MouseLocal::turnMouseLocal(int halfStepsLeft, int halfStepsRight) {
    try {
        const int numPossibleDirections = Constants::MouseConstants::possibleMouseDirections.size();
        const auto& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;

        int currentIndex = findDirectionIndexInPossibleDirections(mouseDirection);
        int newIndex = (currentIndex + halfStepsRight - halfStepsLeft + numPossibleDirections) % numPossibleDirections;
        mouseDirection = possibleMouseDirections[newIndex];
    }
    catch(const std::invalid_argument& e) {
        std::cerr << e.what() << std::endl;
        // Handle exception as needed
    }
}

// Adjusts the mouse's direction based on a desired direction
std::array<int, 2> MouseLocal::turnMouseLocal(const std::array<int, 2>& newDirection) {
    std::array<int, 2> halfStepCount = obtainHalfStepCount(newDirection);
    int halfStepsLeft = (halfStepCount[1] == -1) ? halfStepCount[0] : 0;
    int halfStepsRight = (halfStepCount[1] == 1) ? halfStepCount[0] : 0;
    turnMouseLocal(halfStepsLeft, halfStepsRight);
    return halfStepCount;
}

// Obtains optimal number of half steps to reach a desired direction
std::array<int, 2> MouseLocal::obtainHalfStepCount(const std::array<int, 2>& newDirection) {
    const auto& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    int numDirections = possibleMouseDirections.size();

    int currentIndex = findDirectionIndexInPossibleDirections(mouseDirection);
    int newDirIndex = findDirectionIndexInPossibleDirections(newDirection);

    int halfStepsRight = (newDirIndex - currentIndex + numDirections) % numDirections;
    int halfStepsLeft = (currentIndex - newDirIndex + numDirections) % numDirections;

    int halfSteps = std::min(halfStepsRight, halfStepsLeft);
    int direction = (halfStepsRight < halfStepsLeft) ? 1 : -1; // 1 for right, -1 for left

    return { halfSteps, direction };
}

// Finds the index of a given direction
int MouseLocal::findDirectionIndexInPossibleDirections(const std::array<int, 2>& direction) const {
    const auto& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    for(size_t i = 0; i < possibleMouseDirections.size(); ++i) {
        if(possibleMouseDirections[i][0] == direction[0] && possibleMouseDirections[i][1] == direction[1]) {
            return i;
        }
    }
    throw std::invalid_argument("Direction not listed as a possible mouse direction.");
}

// Moves the mouse forward in the current direction
void MouseLocal::moveForwardLocal() {
    int newXPosition = mousePosition[0] + mouseDirection[0];
    int newYPosition = mousePosition[1] + mouseDirection[1];

    if(isValidCell(newXPosition, newYPosition)) {
        mousePosition = { newXPosition, newYPosition };
    }
    else {
        std::cerr << "Invalid position (ack), mouse cannot move to (" << newXPosition << "," << newYPosition << ")" << std::endl;
    }
}

// Adds a wall locally
void MouseLocal::addWallLocal(int x, int y, const std::array<int, 2>& direction) {
    int neighboringCellX = x + direction[0];
    int neighboringCellY = y + direction[1];

    if(isValidCell(neighboringCellX, neighboringCellY)) {
        mazeCells[y][x]->addWall(direction, true);
        mazeCells[neighboringCellY][neighboringCellX]->addWall({ -direction[0], -direction[1] }, true);
        // std::cerr << "Shared wall cell found :)" << std::endl;
    }
    else {
        mazeCells[y][x]->addWall(direction, false);
        // std::cerr << "Edge cell found :)" << std::endl; // FIXME: Remove later.
    }
}

// Checks if cell is valid
bool MouseLocal::isValidCell(int x, int y) const {
    return x >= 0 && x < Constants::MazeConstants::numCols && y >= 0 && y < Constants::MazeConstants::numRows;
}

// Determines movement between two cells
Movement MouseLocal::getMovement(const Cell& cell1, const Cell& cell2, bool diagonalsAllowed) {
    std::array<int, 2> direction = { cell2.getX() - cell1.getX(), cell2.getY() - cell1.getY() };
    try {
        double distance = std::sqrt(std::pow(direction[0], 2) + std::pow(direction[1], 2));
        if(distance != 1 && distance != std::sqrt(2)) {
            return Movement(false, direction);
        }

        // Cardinal direction movement
        bool canMove = !cell1.getWallExists(direction);
        Movement cardinalMovement(canMove, direction);

        return cardinalMovement;
    }
    catch(const std::invalid_argument& e) {
        Movement diagonalMovement(false, direction);
        if(!diagonalsAllowed || !cell2.getIsExplored()) {
            return diagonalMovement;
        }

        // For diagonals, check if both horizontal and vertical directions are clear
        std::array<int, 2> horizontalDirectionCheck = { direction[0], 0 };
        std::array<int, 2> verticalDirectionCheck = { 0, direction[1] };

        bool diagonalHorizontal = (!cell1.getWallExists(horizontalDirectionCheck) &&
                                    !cell2.getWallExists({ 0, -verticalDirectionCheck[1] }));
        bool diagonalVertical = (!cell1.getWallExists(verticalDirectionCheck) &&
                                  !cell2.getWallExists({ -horizontalDirectionCheck[0], 0 }));

        diagonalMovement.setCanMove(diagonalHorizontal || diagonalVertical);

        if(!diagonalMovement.getCanMove()) {
            return diagonalMovement;
        }

        // Determine if the diagonal is left or right
        bool isHorizontalRight = (findDirectionIndexInPossibleDirections(verticalDirectionCheck) <
                                   findDirectionIndexInPossibleDirections(horizontalDirectionCheck));
        bool isVerticalRight = !isHorizontalRight;

        if(!isHorizontalRight && diagonalHorizontal) {
            diagonalMovement.setIsDiagonal(true);
            diagonalMovement.setLeftOrRightDiagonal("left");
            diagonalMovement.setCellToMoveToFirst(&getCell(cell1.getX() + horizontalDirectionCheck[0],
                                                          cell1.getY()));
        }
        else if(!isVerticalRight && diagonalVertical) {
            diagonalMovement.setIsDiagonal(true);
            diagonalMovement.setLeftOrRightDiagonal("left");
            diagonalMovement.setCellToMoveToFirst(&getCell(cell1.getX(),
                                                          cell1.getY() + verticalDirectionCheck[1]));
        }
        else if(isHorizontalRight && diagonalHorizontal) {
            diagonalMovement.setIsDiagonal(true);
            diagonalMovement.setLeftOrRightDiagonal("right");
            diagonalMovement.setCellToMoveToFirst(&getCell(cell1.getX() + horizontalDirectionCheck[0],
                                                          cell1.getY()));
        }
        else if(isVerticalRight && diagonalVertical) {
            diagonalMovement.setIsDiagonal(true);
            diagonalMovement.setLeftOrRightDiagonal("right");
            diagonalMovement.setCellToMoveToFirst(&getCell(cell1.getX(),
                                                          cell1.getY() + verticalDirectionCheck[1]));
        }
        else {
            diagonalMovement.setCanMove(false);
        }

        return diagonalMovement;
    }
}

// Returns direction offset based on string
std::array<int, 2> MouseLocal::getDirectionOffset(const std::string& direction) const {
    const std::vector<std::array<int,2>>& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    std::vector<std::string> possibleDirections = { "n", "ne", "e", "se", "s", "sw", "w", "nw" };
    auto it = std::find(possibleDirections.begin(), possibleDirections.end(), direction);
    if(it != possibleDirections.end()) {
        size_t index = std::distance(possibleDirections.begin(), it);
        return possibleMouseDirections[index];
    }
    else {
        throw std::invalid_argument("Invalid direction string: " + direction);
    }
}

// Returns direction as string based on direction offset
std::string MouseLocal::getDirectionAsString(const std::array<int, 2>& direction) const {
    const std::vector<std::string> possibleDirections = { "n", "ne", "e", "se", "s", "sw", "w", "nw" };
    const std::vector<std::array<int,2>>& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    for(size_t i=0; i < possibleMouseDirections.size(); ++i) {
        if(possibleMouseDirections[i][0] == direction[0] && possibleMouseDirections[i][1] == direction[1]) {
            return possibleDirections[i];
        }
    }
    throw std::invalid_argument("Invalid direction offset.");
}

// Returns direction to the left
std::string MouseLocal::getDirectionToTheLeft() const {
    const std::vector<std::array<int,2>>& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    size_t currentIndex = findDirectionIndexInPossibleDirections(mouseDirection);
    size_t leftIndex = (currentIndex + 6) % possibleMouseDirections.size(); // Equivalent to turning left
    std::array<int, 2> newDirection = possibleMouseDirections[leftIndex];
    return getDirectionAsString(newDirection);
}

// Returns direction to the right
std::string MouseLocal::getDirectionToTheRight() const {
    const std::vector<std::array<int,2>>& possibleMouseDirections = Constants::MouseConstants::possibleMouseDirections;
    size_t currentIndex = findDirectionIndexInPossibleDirections(mouseDirection);
    size_t rightIndex = (currentIndex + 2) % possibleMouseDirections.size(); // Equivalent to turning right
    std::array<int, 2> newDirection = possibleMouseDirections[rightIndex];
    return getDirectionAsString(newDirection);
}

// Returns mouse direction
std::array<int, 2> MouseLocal::getMouseDirection() const {
    return mouseDirection;
}

// Sets mouse direction
void MouseLocal::setMouseDirection(const std::array<int, 2>& newDirection) {
    mouseDirection = newDirection;
}

// Returns the maze cells
const std::vector<std::vector<Cell*>>& MouseLocal::getMazeCells() const {
    return mazeCells;
}

// Returns a particular cell
Cell& MouseLocal::getCell(int x, int y) {
    if(!isValidCell(x, y)) {
        throw std::out_of_range("Invalid cell coordinates.");
    }
    return *mazeCells[y][x];
}

Cell& MouseLocal::getCell(int x, int y) const {
    if(!isValidCell(x, y)) {
        throw std::out_of_range("Invalid cell coordinates.");
    }
    return *mazeCells[y][x];
}

// Returns mouse's current position cell
Cell& MouseLocal::getMousePosition() {
    return getCell(mousePosition[0], mousePosition[1]);
}

// Sets mouse's position
void MouseLocal::setMousePosition(const Cell& newMousePosition) {
    mousePosition = { newMousePosition.getX(), newMousePosition.getY() };
}

// Returns a string representation of the local maze
std::string MouseLocal::localMazeToString() const {
    std::stringstream mazeString;
    mazeString << "Maze:\n";
    int numRows = Constants::MazeConstants::numRows;
    int numCols = Constants::MazeConstants::numCols;

    // Print top boundary
    for(int i = 0; i < numCols; ++i) {
        mazeString << "+---";
    }
    mazeString << "+\n";

    for(int i = numRows - 1; i >=0; --i) {
        mazeString << printRow(i);
        mazeString << "\n";
    }

    return mazeString.str();
}

// Helper method to print a row
std::string MouseLocal::printRow(int rowNumber) const {
    std::stringstream rowString;
    rowString << "|";

    int numCols = Constants::MazeConstants::numCols;

    for(int i = 0; i < numCols; ++i) {
        const Cell& cellAnalyzed = getCell(i, rowNumber);
        if(cellAnalyzed.getWallExists({1,0})) {
            rowString << "   |";
        }
        else {
            rowString << "    ";
        }
    }
    rowString << "\n";

    for(int i = 0; i < numCols; ++i) {
        const Cell& cellAnalyzed = getCell(i, rowNumber);
        if(cellAnalyzed.getWallExists({0,-1})) {
            rowString << "+---";
        }
        else {
            rowString << "+   ";
        }
    }
    rowString << "+\n";

    return rowString.str();
}

// Get neighbors
std::vector<Cell*> MouseLocal::getNeighbors(const Cell& cell, bool diagonalsAllowed) const {
    int x = cell.getX();
    int y = cell.getY();
    std::vector<Cell*> neighbors;

    const std::vector<std::array<int, 2>>& possibleDirections = Constants::MouseConstants::possibleMouseDirections;

    for(const auto& direction : possibleDirections) {
        int newX = x + direction[0];
        int newY = y + direction[1];
        if(isValidCell(newX, newY)) {
            if(diagonalsAllowed || direction[0] == 0 || direction[1] == 0) {
                //.. neighbors.emplace_back(getCell(newX, newY));
                //.. neighbors.emplace_back(new Cell(getCell(newX, newY)));
            }
        }
    }
    return neighbors;
}

// Static method to check if two cells are the same
bool MouseLocal::isSame(const Cell& cell1, const Cell& cell2) {
    return cell1.getX() == cell2.getX() && cell1.getY() == cell2.getY();
}

// Static method to calculate Euclidean distance
double MouseLocal::euclideanDistance(const Cell& cell1, const Cell& cell2) {
    return std::sqrt(std::pow(cell1.getX() - cell2.getX(), 2) + std::pow(cell1.getY() - cell2.getY(), 2));
}

// Static method to calculate octile distance
double MouseLocal::octileDistance(const Cell& cell1, const Cell& cell2) {
    int distanceX = std::abs(cell1.getX() - cell2.getX());
    int distanceY = std::abs(cell1.getY() - cell2.getY());
    return (distanceX + distanceY) + (std::sqrt(2) - 2) * std::min(distanceX, distanceY);
}

// Resets costs of all cells
void MouseLocal::resetCosts() {
    for(int x = 0; x < Constants::MazeConstants::numCols; ++x) {
        for(int y = 0; y < Constants::MazeConstants::numRows; ++y) {
            Cell& cell = getCell(x, y);
            cell.setCostFromStart(std::numeric_limits<double>::infinity());
            cell.setTotalCost(std::numeric_limits<double>::infinity());
        }
    }
}

// Detects and sets walls using the API
void MouseLocal::detectAndSetWalls(API& api) {
    Cell& currCell = getMousePosition();
    if(api.wallFront()) {
        api.setWall(currCell.getX(), currCell.getY(), getDirectionAsString(getMouseDirection()));
    }
    if(api.wallLeft()) {
        api.setWall(currCell.getX(), currCell.getY(), getDirectionToTheLeft());
    }
    if(api.wallRight()) {
        api.setWall(currCell.getX(), currCell.getY(), getDirectionToTheRight());
    }
}

// Retrieves goal cells from Constants
std::vector<Cell*> MouseLocal::getGoalCells() const {
    std::vector<std::array<int,2>> goalPoses = Constants::MazeConstants::getGoalCells();
    std::vector<Cell*> goalCells;
    for(const auto& goalPos : goalPoses) {
        goalCells.emplace_back(&getCell(goalPos[0], goalPos[1]));
    }
    return goalCells;
}

// Checks if a cell is a goal cell
bool MouseLocal::isGoalCell(const Cell& cell, const std::vector<Cell*>& goalCells) const {
    bool isGoal = false;
    for(const auto& goal : goalCells) {
        isGoal = isGoal || isSame(cell, *goal);
    }
    return isGoal;
}

// Gets direction between two cells
std::array<int, 2> MouseLocal::getDirBetweenCells(const Cell& cell1, const Cell& cell2) {
    return { cell2.getX() - cell1.getX(), cell2.getY() - cell1.getY() };
}